tests/contractFT_test.cpp

```diff
@@ -0,0 +1,264 @@
+{
+
+  ColorSpinorParam cs_param;
+
+  cs_param.nColor = 3;
+  cs_param.nSpin = nSpin;
+  cs_param.nDim = 4;
+
+  for (int i = 0; i < 4; i++) cs_param.x[i] = X[i];
+
+  cs_param.x[4] = 1;
+  cs_param.siteSubset = QUDA_FULL_SITE_SUBSET;
+  cs_param.setPrecision(sizeof(Float) == sizeof(float) ? QUDA_SINGLE_PRECISION : QUDA_DOUBLE_PRECISION);
+
+  cs_param.pad = 0;
+  cs_param.siteOrder = QUDA_EVEN_ODD_SITE_ORDER;
+  cs_param.fieldOrder = QUDA_SPACE_SPIN_COLOR_FIELD_ORDER;
+  cs_param.gammaBasis = QUDA_DEGRAND_ROSSI_GAMMA_BASIS; // meaningless for staggered, but required by the code.
+  cs_param.create = QUDA_ZERO_FIELD_CREATE;
+  cs_param.location = QUDA_CPU_FIELD_LOCATION;
+  cs_param.pc_type = QUDA_4D_PC;
+
+  const int my_spinor_site_size = nSpin * 3; // nSpin X nColor
+
+  const int spinor_field_floats = V * my_spinor_site_size * 2; // DMH: Vol * spinor elems * 2(re,im)
+
+  const int n_contract_results = red_size * n_mom * nSpin * nSpin * 2;
+
+  std::vector<double> d_result(n_contract_results, 0.0);
+
+  constexpr int nprops = nSpin * src_colors;
+
+  const int dof = my_spinor_site_size * 2 * nprops;
+
+  // array of spinor field for each source spin and color
+  size_t off = 0;
+
+  // array of spinor field for each source spin and color
+  std::array<void *, nprops> spinorX;
+  std::array<void *, nprops> spinorY;
+
+  std::array<std::vector<Float>, 2> buffs {std::vector<Float>(nprops * spinor_field_floats, 0),
+                                           std::vector<Float>(nprops * spinor_field_floats, 0)};
+  //
+  fill_buffers<Float, 2>(buffs, X, dof);
+  //
+  for (int s = 0; s < nprops; ++s, off += spinor_field_floats * sizeof(Float)) {
+    spinorX[s] = (void *)((uintptr_t)buffs[0].data() + off);
+    spinorY[s] = (void *)((uintptr_t)buffs[1].data() + off);
+  }
+  // Perform GPU contraction:
+  void *d_result_ = static_cast<void *>(d_result.data());
+
+  contractFTQuda(spinorX.data(), spinorY.data(), &d_result_, cType, (void *)(&cs_param), src_colors, X.data(),
+                 source_position.data(), n_mom, mom.data(), fft_type.data());
+  // Check results:
+  int faults
+    = contractionFT_reference<Float>((Float **)spinorX.data(), (Float **)spinorY.data(), d_result.data(), cType,
+                                     src_colors, X.data(), source_position.data(), n_mom, mom.data(), fft_type.data());
+
+  return faults;
+}
+
+template <typename Float, int src_colors, int n_mom>
+int launch_contract_test(const QudaContractType cType, const std::array<int, 4> &X, const int nspin, const int red_size,
+                         const std::array<int, 4> &source_position, const std::array<int, n_mom * 4> &mom,
+                         const std::array<QudaFFTSymmType, n_mom * 4> &fft_type)
+{
+  int faults = 0;
+
+  if (nspin == 1) {
+    faults = launch_contract_test<Float, 1, src_colors, n_mom>(cType, X, red_size, source_position, mom, fft_type);
+    //} else  if ( nspin == 4 ){
```

Comment:
Should this be un-commented?

Reply:
No, since spin 4 case was not tested (activated) in this branch. We agreed that this can be done later, but I added a comment (TODO message)